core/src/reth/exex.rs

```diff
@@ -0,0 +1,338 @@
+
+use crate::reth::types::{
+    BlockRangeInclusiveIter, ExExMode, ExExRequest, ExExReturnData, LogMetadata,
+};
+/// The default parallelism for the ExEx backfill jobs.
+const DEFAULT_PARALLELISM: usize = 32;
+/// The default batch size for the ExEx backfill jobs.
+const DEFAULT_BATCH_SIZE: usize = 8;
+/// The maximum number of headers we read at once when handling a range filter.
+const MAX_HEADERS_RANGE: u64 = 1_000; // with ~530bytes per header this is ~500kb
+/// The number of concurrent backfills allowed.
+const BACKFILL_CONCURRENCY: usize = 32;
+
+/// State for each job
+struct JobState {
+    filter: Arc<Filter>,
+    tx: mpsc::UnboundedSender<ExExReturnData>,
+    backfill_running: bool,
+    buffer: Mutex<Chain>,
+    handle: Option<tokio::task::JoinHandle<()>>,
+}
+
+/// The ExEx that consumes new [`ExExNotification`]s and processes new backfill requests.
+pub(crate) struct RindexerExEx<Node: FullNodeComponents> {
+    /// The semaphore that limits the number of concurrent backfills.
+    backfill_semaphore: Arc<Semaphore>,
+    /// The context of the ExEx.
+    ctx: ExExContext<Node>,
+    /// Sender for exex requests.
+    request_tx: mpsc::UnboundedSender<ExExRequest>,
+    /// Receiver for exex requests.
+    request_rx: mpsc::UnboundedReceiver<ExExRequest>,
+    /// jobs
+    jobs: HashMap<u64, JobState>,
+    /// The next job id
+    next_job_id: u64,
+}
+
+impl<Node> RindexerExEx<Node>
+where
+    Node: FullNodeComponents<Types: NodeTypes<Primitives = EthPrimitives>>,
+{
+    /// Creates a new instance of the ExEx.
+    pub(crate) fn new(
+        ctx: ExExContext<Node>,
+        request_tx: mpsc::UnboundedSender<ExExRequest>,
+        request_rx: mpsc::UnboundedReceiver<ExExRequest>,
+    ) -> Self {
+        Self {
+            ctx,
+            request_tx,
+            request_rx,
+            next_job_id: 0,
+            jobs: HashMap::new(),
+            backfill_semaphore: Arc::new(Semaphore::new(BACKFILL_CONCURRENCY)),
+        }
+    }
+
+    /// Starts listening for notifications and backfill requests.
+    pub(crate) async fn start(mut self) -> eyre::Result<()> {
+        loop {
+            tokio::select! {
+                Some(notification) = self.ctx.notifications.next() => {
+                    self.handle_notification(notification?).await?;
+                }
+                Some(message) = self.request_rx.recv() => {
+                    self.handle_request(message).await;
+                }
+            }
+        }
+    }
+
+    /// Handles the given notification and calls [`process_committed_chain`] for a committed
+    /// chain, if any.
+    async fn handle_notification(&self, notification: ExExNotification) -> eyre::Result<()> {
+        match &notification {
+            ExExNotification::ChainCommitted { new } => {
+                info!(committed_chain = ?new.range(), "Received commit");
+            }
+            ExExNotification::ChainReorged { old, new } => {
```

Comment:
we should do reorgs later in the whole rindexer project

Reply:
Yes! Kept some empty logs here to make sure it's extendable. But I'll add a TODO comment here as well.